--- /home/axel/build/llvm-orig/src/./tools/clang/include/clang/Basic/SourceManager.h	2013-04-11 09:35:12.700970247 +0200
+++ ./tools/clang/include/clang/Basic/SourceManager.h	2013-04-24 19:40:12.173515046 +0200
@@ -1456,7 +1456,7 @@
       if (Invalid) *Invalid = true;
       return LocalSLocEntryTable[0];
     }
-    return getSLocEntryByID(FID.ID);
+    return getSLocEntryByID(FID.ID, Invalid);
   }
 
   unsigned getNextLocalOffset() const { return NextLocalOffset; }
@@ -1520,11 +1520,11 @@
   const SrcMgr::SLocEntry &loadSLocEntry(unsigned Index, bool *Invalid) const;
 
   /// \brief Get the entry with the given unwrapped FileID.
-  const SrcMgr::SLocEntry &getSLocEntryByID(int ID) const {
+  const SrcMgr::SLocEntry &getSLocEntryByID(int ID, bool* Invalid = 0) const {
     assert(ID != -1 && "Using FileID sentinel value");
     if (ID < 0)
-      return getLoadedSLocEntryByID(ID);
-    return getLocalSLocEntry(static_cast<unsigned>(ID));
+      return getLoadedSLocEntryByID(ID, Invalid);
+    return getLocalSLocEntry(static_cast<unsigned>(ID), Invalid);
   }
 
   const SrcMgr::SLocEntry &getLoadedSLocEntryByID(int ID,
--- /home/axel/build/llvm-orig/src/./tools/clang/lib/Serialization/ASTReader.cpp	2013-04-11 09:34:53.569169094 +0200
+++ ./tools/clang/lib/Serialization/ASTReader.cpp	2013-04-24 19:40:15.673487393 +0200
@@ -932,7 +932,7 @@
     // We will detect whether a file changed and return 'Failure' for it, but
     // we will also try to fail gracefully by setting up the SLocEntry.
     unsigned InputID = Record[4];
-    InputFile IF = getInputFile(*F, InputID);
+    InputFile IF = getInputFile(*F, InputID, /*Complain=*/false);
     const FileEntry *File = IF.getFile();
     bool OverriddenBuffer = IF.isOverridden();
 
@@ -1683,7 +1683,8 @@
         ErrorStr += "' referenced by AST file";
         Error(ErrorStr.c_str());
       }
-      return InputFile();
+      File = FileMgr.getVirtualFile(Filename, StoredSize, StoredTime);
+      //return InputFile();
     }
 
     // Check if there was a request to override the contents of the file
