--- /home/axel/build/cling/src/./lib/Support/DynamicLibrary.cpp	2013-04-24 21:58:01.339752276 +0200
+++ ./lib/Support/DynamicLibrary.cpp	2013-04-24 19:40:23.073428922 +0200
@@ -137,6 +137,10 @@
   }
 
 #if HAVE_DLFCN_H
+  void* ptr = dlsym(RTLD_DEFAULT, symbolName);
+  if (ptr)
+    return ptr;
+#ifdef R__TOO_SLOW
   // Now search the libraries.
   if (OpenedHandles) {
     for (DenseSet<void *>::iterator I = OpenedHandles->begin(),
@@ -148,6 +152,7 @@
       }
     }
   }
+#endif // R__TOO_SLOW
 #endif
 
   if (void *Result = llvm::SearchForAddressOfSpecialSymbol(symbolName))
