--- /home/axel/build/cling/src/./tools/clang/lib/Sema/Sema.cpp	2013-04-24 22:04:42.288514584 +0200
+++ ./tools/clang/lib/Sema/Sema.cpp	2013-04-24 19:40:14.329498012 +0200
@@ -599,7 +599,8 @@
 
   if (TUKind == TU_Prefix) {
     // Translation unit prefixes don't need any of the checking below.
-    TUScope = 0;
+    if (!PP.isIncrementalProcessingEnabled())
+       TUScope = 0;
     return;
   }
 
@@ -790,7 +791,8 @@
   assert(ParsingInitForAutoVars.empty() &&
          "Didn't unmark var as having its initializer parsed");
 
-  TUScope = 0;
+  if (!PP.isIncrementalProcessingEnabled())
+     TUScope = 0;
 }
 
 
