--- /home/axel/build/llvm-orig/src/./tools/clang/lib/Lex/Pragma.cpp	2013-04-11 09:34:54.365160820 +0200
+++ ./tools/clang/lib/Lex/Pragma.cpp	2013-04-24 19:40:14.677495262 +0200
@@ -858,9 +858,11 @@
   }
 
   // Check to make sure we don't already have a pragma for this identifier.
-  assert(!InsertNS->FindHandler(Handler->getName()) &&
-         "Pragma handler already exists for this identifier!");
-  InsertNS->AddPragma(Handler);
+  // assert(!InsertNS->FindHandler(Handler->getName()) &&
+  //        "Pragma handler already exists for this identifier!");
+
+  if (!InsertNS->FindHandler(Handler->getName()))
+     InsertNS->AddPragma(Handler);
 }
 
 /// RemovePragmaHandler - Remove the specific pragma handler from the
