--- ./tools/clang/lib/Driver/ToolChains.cpp	2013-04-11 09:34:54.425160197 +0200
+++ ./tools/clang/lib/Driver/ToolChains.cpp	2013-04-24 19:40:14.813494188 +0200
@@ -1095,6 +1095,7 @@
   static const char *const X86_64Triples[] = {
     "x86_64-linux-gnu", "x86_64-unknown-linux-gnu", "x86_64-pc-linux-gnu",
     "x86_64-redhat-linux6E", "x86_64-redhat-linux", "x86_64-suse-linux",
+    "x86_64-mageia-linux-gnu",
     "x86_64-manbo-linux-gnu", "x86_64-linux-gnu", "x86_64-slackware-linux"
   };
   static const char *const X86LibDirs[] = { "/lib32", "/lib" };
@@ -1102,6 +1103,7 @@
     "i686-linux-gnu", "i686-pc-linux-gnu", "i486-linux-gnu", "i386-linux-gnu",
     "i386-redhat-linux6E", "i686-redhat-linux", "i586-redhat-linux",
     "i386-redhat-linux", "i586-suse-linux", "i486-slackware-linux",
+    "i586-mageia-linux-gnu",
     "i686-montavista-linux"
   };
 
