Index: tools/CMakeLists.txt
===================================================================
--- tools/CMakeLists.txt	(revision 159991)
+++ tools/CMakeLists.txt	(working copy)
@@ -52,4 +52,8 @@
 add_llvm_external_project(lld)
 add_llvm_external_project(polly)
 
+if( EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/cling/CMakeLists.txt )
+  add_llvm_external_project( ${CMAKE_CURRENT_SOURCE_DIR}/cling )
+endif()
+
 set(LLVM_COMMON_DEPENDS ${LLVM_COMMON_DEPENDS} PARENT_SCOPE)
