--- /home/axel/build/cling/src/./lib/Support/PathV2.cpp	2013-04-24 21:58:01.343752243 +0200
+++ ./lib/Support/PathV2.cpp	2013-04-24 19:40:23.077428890 +0200
@@ -884,7 +884,7 @@
 }
 
 error_code identify_magic(const Twine &path, file_magic &result) {
-  SmallString<32> Magic;
+  SmallString<512> Magic;
   error_code ec = get_magic(path, Magic.capacity(), Magic);
   if (ec && ec != errc::value_too_large)
     return ec;
