Index: tools/Makefile
===================================================================
--- tools/Makefile	(revision 159991)
+++ tools/Makefile	(working copy)
@@ -22,6 +22,9 @@
 # Build LLDB if present. Note LLDB must be built last as it depends on the
 # wider LLVM infrastructure (including Clang).
 OPTIONAL_DIRS := lldb
+ifeq ($(NOCLING),)
+OPTIONAL_DIRS += cling
+endif
 
 # NOTE: The tools are organized into five groups of four consisting of one
 # large and three small executables. This is done to minimize memory load
