Index: tools/cling/lib/Interpreter/LookupHelper.cpp
===================================================================
--- tools/cling/lib/Interpreter/LookupHelper.cpp	(revision 46197)
+++ tools/cling/lib/Interpreter/LookupHelper.cpp	(working copy)
@@ -617,7 +617,7 @@
           //  Do template argument deduction and function argument matching.
           //
           FunctionDecl* Specialization;
-          sema::TemplateDeductionInfo TDI(Context, SourceLocation());
+          sema::TemplateDeductionInfo TDI( (SourceLocation()) );
           Sema::TemplateDeductionResult TDR 
             = S.DeduceTemplateArguments(FTD,
                          const_cast<TemplateArgumentListInfo*>(FuncTemplateArgs),
Index: tools/cling/lib/Interpreter/InvocationOptions.cpp
===================================================================
--- tools/cling/lib/Interpreter/InvocationOptions.cpp	(revision 46197)
+++ tools/cling/lib/Interpreter/InvocationOptions.cpp	(working copy)
@@ -25,7 +25,7 @@
   static const OptTable::Info ClingInfoTable[] = {
 #define OPTION(NAME, ID, KIND, GROUP, ALIAS, FLAGS, PARAM,              \
                HELPTEXT, METAVAR)                                       \
-    { NAME, HELPTEXT, METAVAR, Option::KIND##Class, FLAGS, PARAM,       \
+    { NAME, HELPTEXT, METAVAR, OPT_##ID, Option::KIND##Class, FLAGS, PARAM, \
       OPT_##GROUP, OPT_##ALIAS },
 #include "cling/Interpreter/ClingOptions.inc"
 #undef OPTION
